src/main/java/io/cryostat/agent/triggers/TriggerEvaluator.java

```diff
@@ -0,0 +1,210 @@
+                            if (evaluateTriggerConstraint(t, Duration.ZERO)) {
+                                // Condition was met, set the state accordingly
+                                log.trace("Trigger {} satisfied, watching...", t);
+                                t.setState(TriggerState.WAITING_HIGH);
+                                t.setTimeConditionFirstMet(new Date(System.currentTimeMillis()));
+                            } else {
+                                // Condition wasn't met, keep waiting.
+                                log.trace("Trigger {} not yet satisfied...", t);
+                                t.setState(TriggerState.WAITING_LOW);
+                            }
+                        }
+                        break;
+                    case WAITING_HIGH:
+                        // Condition was met at last check but duration hasn't passed
+                        if (handleDuration(difference, t)) {
+                            if (evaluateTriggerConstraint(t, Duration.ofMillis(difference))) {
+                                log.trace("Trigger {} satisfied, completing...", t);
+                                t.setState(TriggerState.COMPLETE);
+                                flightRecorderHelper.startRecording(t.getRecordingTemplateName());
+                            } else {
+                                log.trace("Trigger {} satisfied, waiting for duration...", t);
+                                t.setState(TriggerState.WAITING_LOW);
+                            }
+                        }
+                        break;
+                    case WAITING_LOW:
+                        if (evaluateTriggerConstraint(t, Duration.ofMillis(difference))) {
+                            t.setState(TriggerState.WAITING_HIGH);
+                            t.setTimeConditionFirstMet(new Date(System.currentTimeMillis()));
+                        }
+                        break;
+                }
+            }
+        } catch (Exception e) {
+            log.error("Unexpected exception during evaluation", e);
+        }
+    }
+
+    private boolean handleDuration(long difference, SmartTrigger trigger) {
+        if (trigger.getDurationConstraint().contains("<=")) {
+            return trigger.getTargetDuration().toMillis() <= difference;
+        } else if (trigger.getDurationConstraint().contains(">=")) {
+            return trigger.getTargetDuration().toMillis() >= difference;
+        } else if (trigger.getDurationConstraint().contains(">")) {
+            return trigger.getTargetDuration().toMillis() > difference;
+        } else if (trigger.getDurationConstraint().contains("<")) {
+            return trigger.getTargetDuration().toMillis() < difference;
+        } else {
+            return trigger.getTargetDuration().toMillis() == difference;
+        }
+    }
+
+    private boolean evaluateTriggerConstraint(SmartTrigger trigger, Duration target) {
```

Comment:
@Josh-Matsuoka I'm trying to follow the logic here to troubleshoot the triggers with a duration. I think I'm getting the picture, but it looks like this `target` parameter is not actually used within the method body here - looks like it's probably related to the bug I'm seeing. Should that `target` parameter be passed in to the `scriptVars` map rather than sticking to `trigger.getTargetDuration()`? That way the duration being compared in the CEL expression depends on the state the trigger is in.

Reply:
Yes, that should be what happens, I think that got lost the last time I reworked that method while implementing it, good catch.